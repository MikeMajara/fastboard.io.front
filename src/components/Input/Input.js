import React, { useCallback, forwardRef } from 'react';
import PropTypes from 'prop-types';

import SInput from './styled';

const testId = 'input-component';

const Input = forwardRef(({
  name, value, placeholder, onChange : onChangeProp, fullWidth,
}, ref) => {
  const onChange = useCallback(
    (evt) => onChangeProp(evt.target.value),
    [onChangeProp],
  );

  return (
    <SInput
      ref={ref}
      name={name}
      value={value}
      placeholder={placeholder}
      onChange={onChange}
      fullWidth={fullWidth}
      data-testid={testId}
    />
  );
});

Input.defaultProps = {
  value       : undefined,
  placeholder : undefined,
  fullWidth   : false,
  onChange    : undefined,
};

Input.propTypes = {
  name        : PropTypes.string.isRequired,
  value       : PropTypes.string,
  placeholder : PropTypes.string,
  fullWidth   : PropTypes.bool,
  onChange    : PropTypes.func,
};

export default Input;
