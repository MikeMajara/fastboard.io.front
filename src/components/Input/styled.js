import styled from '@xstyled/styled-components';

export default styled.input`
  width: ${({ fullWidth }) => (fullWidth ? '100%' : 'auto')};
  height: formHeight;
  box-sizing: border-box;
  padding: xs md;
  border-width: default;
  border-style: default;
  border-color: border;
  background-color: bg;
  border-radius: default;
  color: fg;
  outline: none;

  &:focus {
    border-color: primary;
  }

  &::placeholder {
    color: border;
    font-size: 14px;
  }
`;
