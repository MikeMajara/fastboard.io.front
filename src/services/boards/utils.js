/**
 * Send boards event trough the network.
 *
 * @param {String} eventName Event key name.
 * @param {Object} data Data associated to the event.
 * @param {Function} ask Callback for server event acknowledgement.
 */
function send(eventName, data, ack) {
  this.socket.emit(eventName, data, ack);
}

/**
 * Check if the service is initialized.
 *
 * @return {Boolean} True if the service is initialized; false otherwise.
 */
function isInit() {
  return this.isInit;
}

export {
  isInit,
  send,
};
