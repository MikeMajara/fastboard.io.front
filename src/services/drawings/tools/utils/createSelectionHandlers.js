import { Group, Shape } from 'paper';

import { Log } from '#utils';
import { cursorTypes, selectionColorCode } from '#constants';
import store, { setSelectorCursorHover } from '#store';

import { bounds } from '../selector/constants';

const boundsArray = Object.values(bounds);
const bounds2cursorType = Object.freeze({
  [bounds.topLeft]     : cursorTypes.rotating,
  [bounds.topRight]    : cursorTypes.menu,
  [bounds.bottomLeft]  : cursorTypes.resizingSW,
  [bounds.bottomRight] : cursorTypes.resizingSE,
});

/**
 * Add selection handlers to the selected item.
 *
 * @param {Object} selectedItem Item to be added selection handlers to.
 * @param {Object} selectionLayer Layer that will contain the selection handlers.
 *
 * @returns {Object} Item - group with 4 handler items - from the selection handler layer.
 */
export default function createSelectionHandlers(selectedItem, selectionLayer) {
  Log.debug('Service : Drawings : Tools : Utils : createSelectionHandlers', { selectedItem });

  // Set selected item to show its selection state
  selectedItem.set({
    selected      : true,
    selectedColor : selectionColorCode,
    onMouseEnter  : () => store.dispatch(setSelectorCursorHover(cursorTypes.dragging)),
    onMouseLeave  : () => store.dispatch(setSelectorCursorHover()),
  });

  // Create bound handlers on the selection layer
  boundsArray.forEach((bound) => {
    selectionLayer.addChild(new Shape.Circle({
      name         : bound,
      center       : selectedItem.internalBounds[bound],
      radius       : 6,
      fillColor    : selectionColorCode,
      onMouseEnter : () => store.dispatch(setSelectorCursorHover(bounds2cursorType[bound])),
      onMouseLeave : () => store.dispatch(setSelectorCursorHover()),
    }));
  });

  // Create a group with the handlers to manipulate all of them at once
  selectionLayer.addChild(
    new Group({
      name     : 'selectionHandlers', // @todo to constants
      children : boundsArray.map(
        (bound) => selectionLayer.children[bound],
      ),
    }),
  );

  // Reproduce the selected item transformations on the handler items group
  selectionLayer.children.selectionHandlers.transform(selectedItem.matrix);

  // Except the scaling of the handler items themselves, which retain the original size
  selectionLayer.children.selectionHandlers.children.forEach((handlerItem) => {
    handlerItem.set({ scaling : 1 });
  });

  return selectionLayer.children.selectionHandlers;
}
