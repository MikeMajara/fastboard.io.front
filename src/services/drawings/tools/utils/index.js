export { default as createSelectionHandlers } from './createSelectionHandlers';
export { default as removeSelectionHandlers } from './removeSelectionHandlers';
export { default as resizeSelectionHandlers } from './resizeSelectionHandlers';
export { default as rotateSelectionHandlers } from './rotateSelectionHandlers';
export { default as translateSelectionHandlers } from './translateSelectionHandlers';
