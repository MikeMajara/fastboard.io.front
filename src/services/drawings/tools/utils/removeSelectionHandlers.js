import { Log } from '#utils';

/**
 * Remove selection handlers.
 *
 * @param {Object} selectedItem Item to be removed selection handlers from.
 * @param {Object} selectionLayer Layer that contains the selection handlers.
 */
export default function removeSelectionHandlers(selectedItem, selectionLayer) {
  Log.debug('Service : Drawings : Tools : Utils : removeSelectionHandlers');

  selectedItem.set({ selected : false });
  selectionLayer.removeChildren();
}
