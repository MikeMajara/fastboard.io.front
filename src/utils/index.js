export { default as Log } from './logger';
export { default as generateId } from './generateId';
export { default as point2net } from './point2net';
export { default as setPreventUnload } from './setPreventUnload';
export { default as throttle } from './throttle';
export { default as timeoutPromise } from './timeoutPromise';
export { default as triggerDownload } from './triggerDownload';
export { default as triggerUpload } from './triggerUpload';
export { default as validBoardId } from './validBoardId';
