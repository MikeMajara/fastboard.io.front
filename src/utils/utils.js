/**
 * Throttles a function call.
 *
 * @param {Function} callback Function to be throttled.
 * @param {Integer} delay Throttle delay in milliseconds.
 *
 * @return {Function} Function that throttles the execution of the callback.
 */
export const throttle = (callback, delay = 0) => {
  let previousTime = new Date().getTime();

  return function throttledCallback(args) {
    const currentTime = new Date().getTime();

    if ((currentTime - previousTime) >= delay) {
      previousTime = currentTime;
      callback(...args);
    }
  };
};

/**
 * Creates a Promise with a rejection timeout.
 *
 * @param {Function} callback Function for the promise. Receives res and rej functions.
 * @param {Integer} timeout Timeout for the promise to be rejected in milliseconds.
 *
 * @return {Promise} Promise with rejection timeout.
 */
export const timeoutPromise = (callback, timeout) => new Promise(
  (res, rej) => {
    const timeoutId = setTimeout(() => rej(new Error()), timeout);

    callback(
      (...args) => {
        clearTimeout(timeoutId);
        res(...args);
      },
      (...args) => {
        clearTimeout(timeoutId);
        rej(new Error(...args));
      },
    );
  },
);
