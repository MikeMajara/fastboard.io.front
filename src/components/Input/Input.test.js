
import React from 'react';
import { fireEvent } from '@testing-library/react';

import { render } from '#devTools/tests/utils';

import Input from './Input';

const originalText = 'some original text';
const changedText = 'some changed text';

describe('Component : Input', () => {
  test('Is visible and reacts to changes', () => {
    const onChangeFn = jest.fn();
    const { queryByTestId } = render(
      <Input
        name="sample"
        value={originalText}
        onChange={onChangeFn}
      />,
    );

    const input = queryByTestId('input-component');

    expect(input).toBeInTheDocument();
    expect(input.value).toBe(originalText);
    expect(onChangeFn).toHaveBeenCalledTimes(0);

    fireEvent.change(input, { target : { value : changedText } });

    expect(onChangeFn).toHaveBeenCalledTimes(1);
    expect(onChangeFn).toHaveBeenCalledWith(changedText);
  });

  test('Snapshot', () => {
    const { asFragment } = render(
      <Input
        name="sample"
        value={originalText}
        onChange={jest.fn()}
      />,
    );

    const componentRender = asFragment();
    expect(componentRender).toMatchSnapshot();
  });
});
