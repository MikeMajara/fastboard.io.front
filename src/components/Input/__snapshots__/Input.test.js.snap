// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component : Input Snapshot 1`] = `
<DocumentFragment>
  <input
    class="sc-pscky hhRpgo"
    data-testid="input-component"
    name="sample"
    value="some original text"
  />
</DocumentFragment>
`;
